File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/SearchMatrix.java
package com.only4.algorithm.version4j.leetcode;

/**
 * @author zhenyu.jiang
 */
public class SearchMatrix {
    public boolean searchMatrix(int[][] matrix, int target) {
        int row = 0;
        int col = matrix[0].length - 1;

        while (row < matrix.length && col >= 0) {
            int current = matrix[row][col];
            if (current == target) {
                return true;
            } else if (current > target) {
                col--;
            } else {
                row++;
            }
        }
        return false;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/searchMatrix.kt
package com.only4.algorithm.version4k.leetcode

fun searchMatrix(matrix: Array<IntArray>, target: Int): Boolean {
    if (matrix.isEmpty() || matrix[0].isEmpty()) return false

    val rows = matrix.size
    val cols = matrix[0].size

    var row = 0
    var col = cols - 1

    while (row < rows && col >= 0) {
        when {
            matrix[row][col] == target -> return true
            matrix[row][col] > target -> col--
            else -> row++
        }
    }
    return false
}
